### DISCUSSION TITLE
Help getting preview in VS code to work on WSL2

### DISCUSSION THREAD

**Discussion by Emilyaoc** [quarto-dev/quarto-cli#7572] [Q&A]
### Description

I have installed quarto within WSL2. I use VS code. When I try to use quarto on the command line within the terminal window in VS code it seems to work ok (i.e. it renders a pdf from example code given in the 'getting started' tutorial on the quarto website). But when I try to use quarto in the file view window e.g. to press preview and see a rendered pdf i get an error message saying "Quarto installation not found". So it seems it's not finding the quarto installation within WSL2 when I try to use it in the viewing window. I looked through the discussion forum and found some posts about needing to specify the path to the quarto CLI executable in the extensions settings within VS code. But I'm quite new to working on the command line & VS code (sorry!) and not 100% sure where the executable for quarto is on my computer. I found an executable file called 'quarto' within /usr/local/bin/. I tried specifying this as the path to the Quarto CLI execuatble as 
"quarto.path": "/opt/quarto/bin/quarto" 
but that gives the error message "Unable to find specified quarto execuatble". Any advice or thoughts would be really welcome! Thank you.

**Comment by cderv** [comment]
> I have installed quarto within WSL2.

Which Linux distribution are you running ? 

> I found an executable file called 'quarto' within /usr/local/bin/

When you do `which quarto` in your WSL2 terminal (not inside VS CODE), what is the path return ? 

> `/opt/quarto/bin/quarto`

Does this binary works in terminal ? 
````
/opt/quarto/bin/quarto --version 
````

**Reply by Emilyaoc** [reply]
My linux distribution is Ubuntu 18.04.6 LTS.

When I do 'which quarto' in my Ubuntu terminal window I get:
/usr/local/bin/quarto

Yes, that binary works in the terminal as when I run it I get: 
"1.3.450"
But note that I get the same if I just do 'quarto --version' in any of my directories (so I guess that is specified somewhere in my $PATH?).

Thank you for your help!

**Reply by cderv** [reply]
> so I guess that is specified somewhere in my $PATH?

Yes `which quarto` will return what is found in your PATH. 

Now if you do that `which quarto` in VSCODE terminal. Does it return the same path ? 

What happens when you try to unset `quarto.path` from all your settings ? (User, Remote, Workspace and/or Folder) ? 
You can do CTRL + MAJ + P, type settings to open the settings window, and then search for `quarto.path`. 

By default, it should use the version found in your PATH when `quarto.path` is unset

**Comment by Emilyaoc** [comment]
Yes, I get the same path when I run 'which quatro' in VSCODE.

The quarto.path box is blank within the settings. Above the blank box it says "A path to the Quarto CLI executable. By default, the extension looks for Quarto CLI in the PATH, but if set, will use the path specified instead." And the box under this is blank. So it should already be looking for the executable using the PATH if I understand correctly? 

Is it possible that the .qmd file I have open in VSCODE is not 'looking' in my WSL2 side of my computer? Although the .qmd file is saved in a folder within the WSL2 part and that is also where Quarto is installed. But could it somehow be trying to run the code on the windows side of my computer? If that makes any sense?!

**Reply by cderv** [reply]
> And the box under this is blank. So it should already be looking for the executable using the PATH if I understand correctly?

Yes exactly. VSCODE has several layers of settings though, so you should check that all box are empty, but usually it will tell you 
![image](https://github.com/quarto-dev/quarto-cli/assets/6791940/d5de6043-fa3a-4b04-8cbf-0669ca6d56dc)

I don't manage to reproduce yet. Are you using latest version of Quarto extension ?

**Reply by Emilyaoc** [reply]
The extension version is 1.106.4 I think. When I checked what version it was in the extensions menus I say there was a message that said:

"Unknown language in `contributes.grammars.language`. Provided value: typst"

If that sheds any light?

**Reply by cderv** [reply]
> The extension version is 1.106.4 I think.

This was the latest until today. Latest is now 1.107.0 from today : https://marketplace.visualstudio.com/items?itemName=quarto.quarto

Not sure it will solve this issue. 

> "Unknown language in contributes.grammars.language. Provided value: typst"

Where do you see this ? We do define the filed in the extension, but I don't see the error

**Comment by Emilyaoc** [comment]
Yes, the problem remains still with the updated version of the quarto extension. 

I saw the message "Unknown language in contributes.grammars.language. Provided value: typst" when I hovered my cursor over the quarto extension in the extensions marketplace. But now that message seems to have disappared. 

I think from some more digging the problem may be that VS Code is by default looking on the windows side of my computer for everything (and not the WSL2 side). I notice that when I try to run the test script 'hello.qmd' from the 'getting started' tutorial on the quarto website it fails and and a python error "ModuleNotFoundError: No module named 'numpy'". I can see at the top of the interactive screen that it has found a version of python in the windows side of my computer (and is thus not using the version on my WSL2 side). Is there some way I can force VS Code to look for programs etc on the WSL2 side of my computer?

**Reply by cderv** [reply]
> Is there some way I can force VS Code to look for programs etc on the WSL2 side of my computer?

Are you using VSCODE version installed on your WSL2 instance ? or opening the one one from your windows side ? 
Some details at: https://code.visualstudio.com/blogs/2019/09/03/wsl2

By default, it should be the case. 

You can check your PATH maybe ?  `echo $PATH` and see what is inside. Maybe the windows versions of your programs are found in the PATH before any Linux version. But `which quarto` does return the right thing I believe from what you shared. 

> I can see at the top of the interactive screen that it has found a version of python in the windows side of my computer

Maybe this why Python is found there. Check the PATH also. 

I know I have a special config on my WSL2 to not have my Windows PATH gets in the way of my Linux PATH. I wonder if that could be why you get this feeling of things from Windows getting found by default ... 🤔 



<details>
<summary>more details on that </summary>

See `appendWindowsPath` at https://learn.microsoft.com/en-us/windows/wsl/wsl-config#interop-settings

However setting this to `false` has some drawbacks because you will need to manage everything yourself, including making VSCODE from windows side part of the PATH
````sh
 # VSCODE in PATH
vscode_path="/mnt/c/Users/chris/AppData/Local/Programs/Microsoft VS Code/bin"
if [ -d "$vscode_path" ] ; then
    PATH="$vscode_path:$PATH"
 fi
````

</details>


But anyway, setting VSCODE configuration should take precedence over any values from your PATH - so if you configure quarto, it should work. Same for python interpreter.

**Reply by Emilyaoc** [reply]
VSCODE is installed on the windows side of my computer. I use it to work on the WSL2 side of my computer (using the WSL2 extension in VSCODE). 

'echo $PATH' produces what looks to be a series of paths seperated by colons, some paths are within the WSL2 and others within the windows side. Is that a little odd? The first few paths are within the WSL2 though. 

Yes `which quarto` does return a path that confirms that quarto is installed on the WSL2 side. But when I try to use quarto (by pressing the preview button in VSCODE to try and render and view the 'hello.qmd' document) it doesn't work as it seems to be trying to look for python etc on the windows side. This is even though `which python` returns a path on the WSL2 side.

**Reply by cderv** [reply]
> VSCODE is installed on the windows side of my computer. I use it to work on the WSL2 side of my computer (using the WSL2 extension in VSCODE).

Oh ok. I believe I am using VSCODE from within WSL and not just the one from windows with extension. 

This become a VSCODE issue or at least thing to understand - the extension you are using are Windows extension, and I don't know how they react to WSL remote extension. I believe they will continue use config from local windows, unless configured otherwise. 

i'll try to do some test. But in the meantime, you could be able to run VSCODE from withing WSL using 
````powershell
wsl code
````
from your Windows Terminal as described in the online doc I shared in previous post. 

If this opens a new VSCODE windows, it may be a new one, and it may work there (possibly you would need to reinstall some extension if you never used it)

**Reply by Emilyaoc** [reply]
I find I have the same issue regardless of whether I open VS Code from within WSL or connect to WSL as a remote connection. 

I thought I understood that VS Code had to be installed on the windows side of my computer? I got that from this post:
https://code.visualstudio.com/docs/remote/wsl

**Reply by cderv** [reply]
yes it does. You're right that there is two parts. 

So it seems the issue lies elsewhere in configuration. Sory I can't help more right now - I don't have more clues. 

Without reproducing, it is always hard to understand what is happening
